### PR TITLE
[res] addresses unchecked arithmetic overflow / underflow

### PR DESCRIPTION
**Pull request by jorgeantonio21**
Addresses issue #169

### CODE REVIEW

**Review by jorgeantonio21 — COMMENTED** [review]
Currently, the code panics in cases of overflow/underflow. It would be good to handle errors instead, but this involves to change the signatures of some shared methods. More than happy to keep the signatures as they are or refactor accordingly.

**Review by dlubarov — APPROVED** [review]
LGTM otherwise, thanks!

**Comment by dlubarov** [comment]
> Currently, the code panics in cases of overflow/underflow. It would be good to handle errors instead, but this involves to change the signatures of some shared methods. More than happy to keep the signatures as they are or refactor accordingly.

My preference is usually to panic on conditions that are supposed to be impossible. I guess the downside is an unexpected panic could crash a server (though there are workarounds like `catch_unwind`, or auto-restarting with e.g. `monit`), but on the positive side, it keeps the APIs cleaner and more user-friendly (no need for callers to think about how to react to cases that weren't supposed to be possible).

File: mersenne-31/src/lib.rs [hunk 1]
```diff
@@ -266,9 +266,20 @@ impl PrimeField64 for Mersenne31 {
         // In order not to overflow a u64, we must have sum(u) <= 2^32.
         debug_assert!(u.iter().sum::<u64>() <= (1u64 << 32));
 
-        let mut dot = u[0] * v[0].value as u64;
```
[the remark this answers is not in the document]
Comment (jorgeantonio21) [reply mersenne-31/src/lib.rs]: I agree with you. I have rolled back the changes, so we keep it simple and performant :) thanks for the comment

**Comment by jorgeantonio21** [comment]
Addresses #169

**Pull request merged**